Fix #3309 and #3311: Implement Force Network Type Feature

yashraj-01 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #3309. Introduced UI for Force Network Type screen.
Fixes #3311. Implemented logic to force network type.
This PR also fixes #3379. cc: @rt4914

**Mock link:** https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/8ae1a60e-7f6d-4b7b-b741-469bb80007b9/specs/

**Screenshots**

- UI
  <img width="250" height="500" src="https://user-images.githubusercontent.com/63469151/128396893-966b4136-f4d2-4dc1-a33a-7764682198c8.png">

- Espresso tests passing for `DeveloperOptionsFragmentTest`
![Screenshot from 2021-07-31 20-04-10](https://user-images.githubusercontent.com/63469151/127743270-92783027-78e8-40e1-b8fd-4951174ef0cd.png)

- Espresso tests passing for `ForceNetworkTypeActivityTest`
![Screenshot from 2021-07-31 21-16-47](https://user-images.githubusercontent.com/63469151/127745340-08c3c294-05e8-40df-bb9b-742a0826e2af.png)

- Espresso tests passing for `ForceNetworkTypeFragmentTest`
![Screenshot from 2021-07-31 22-10-29](https://user-images.githubusercontent.com/63469151/127750337-6dd06cd9-dd99-4714-b63b-8d4d1bdc5b93.png)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

yashraj-01 committed:
Introduced initial UI for force network type screen

yashraj-01 commented:
I have marked this PR as draft as I have not written the tests as of now. I'm waiting on the approval on PR #3471 and then will follow the same approach to write tests here. Also, the UI needs updates as I have to implement the logic for the selection of network types. That might depend on issue #3310. Thanks.

yashraj-01 committed:
Introduced prod and debug NetworkConnectionUtil

yashraj-01 committed:
Fixed bazel build failure

yashraj-01 committed:
Fixed app layer tests

yashraj-01 committed:
Fixed domain layer tests

yashraj-01 committed:
Fixed testing layer tests

yashraj-01 committed:
Fixed util layer tests

yashraj-01 committed:
Introduced prod and debug tests for NetworkConnectionUtil

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
Nit fixes

yashraj-01 committed:
Renamed files

yashraj-01 committed:
Fixed failing domain layer tests on bazel

yashraj-01 committed:
Fixed failing utility layer tests on Bazel

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
Added named arguments and other fixes

yashraj-01 committed:
Introduced a separate NetworkConnectionTestUtil

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
Fixed failing tests

yashraj-01 committed:
Moved setNetworkInfo to companion object

yashraj-01 committed:
Added test for NetworkConnectionTestUtil

yashraj-01 committed:
Merge branch 'develop' into ui-force-network-type

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
Merge branch 'prod-and-debug-networkconnectionutil' into ui-force-network-type

yashraj-01 committed:
Merge branch 'develop' into ui-force-network-type

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
Changes to BUILD.bazel files and tests. Mostly nits.

yashraj-01 committed:
Moved NetworkConnectionTestUtil and its Test to networking package

yashraj-01 committed:
Renamed debug and prod util

yashraj-01 committed:
Added tests for modules

yashraj-01 committed:
Nit fixes

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
Added missing KDocs

yashraj-01 committed:
Removed redundant kdoc exemption

yashraj-01 committed:
Merge branch 'develop' into ui-force-network-type

yashraj-01 committed:
Implemented force network type feature

yashraj-01 committed:
Merged and resolved merge conflicts

yashraj-01 committed:
Added tests for DeveloperOptionsFragmentTest

yashraj-01 committed:
Introduced ForceNetworkTypeActivityTest and ForceNetworkTypeTestActivity

yashraj-01 committed:
Introduced ForceNetworkTypeFragmentTest

yashraj-01 committed:
Added test file exemptions

yashraj-01 committed:
Added accessibility label exemption

yashraj-01 committed:
Added missing KDocs

yashraj-01 committed:
Bazel nit

yashraj-01 commented:
> I have marked this PR as draft as I have not written the tests as of now. I'm waiting on the approval on PR #3471 and then will follow the same approach to write tests here. Also, the UI needs updates as I have to implement the logic for the selection of network types. That might depend on issue #3310. Thanks.

The tests are now there according to the comments on #3471. The UI is also updated and the logic for the feature is also included as part of this PR only. In short, the PR is now ready for review. PTAL. Thanks,

Sarthak2601 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/ForceNetworkTypeActivityTest.kt:
```diff
@@ -0,0 +1,190 @@
+
+  @Test
+  fun testForceNetworkTypeActivity_forceNetworkTypeFragmentIsDisplayed() {
+    launch(ForceNetworkTypeActivity::class.java).use {
```
This launch command is different from the usually used one in all other test files (eg: MarkTopicsCompletedFragmentTest), any particular reason for this change?

yashraj-01 replied:
Yes. Since the `ForceNetworkTypeActivity` doesn't take any arguments while creating the intent for the activity, we have used the class approach to launch the activity. This was suggested by @anandwana001 in one of my previous PRs.

Sarthak2601 reviewed (commented):
Thanks @yashraj-01. This almost LGTM, just wanted to understand the change in activity launching approach.

Sarthak2601 reviewed (approved):
Good work @yashraj-01, LGTM!

rt4914 commented on app/src/main/res/values/strings.xml:
```diff
@@ -480,4 +481,5 @@
   <string name="developer_options_force_crash">Force crash</string>
   <string name="all">All</string>
   <string name="mark_completed_uppercase">MARK COMPLETED</string>
+  <string name="network_selected">Network is Selected</string>
```
Nit
`   <string name="network_selected">Network is selected</string>`

yashraj-01 replied:
Done.

yashraj-01 replied:
Resolving comment as the nit is fixed.

rt4914 reviewed (approved):
Nice work @yashraj-01

BenHenning commented:
Apologies, will need to look at this a bit later today (my time).

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
Created dedicated local test targets for test/networking

yashraj-01 committed:
Removed extra checks and updated test cases accordingly

yashraj-01 committed:
Migrated tests to networking package

yashraj-01 committed:
Made test util inject

yashraj-01 committed:
Updated test util test

yashraj-01 committed:
Removed kdocs from test files

yashraj-01 committed:
Nit fix

yashraj-01 committed:
Nit fix

yashraj-01 committed:
bazel nit fix

yashraj-01 committed:
Merged and resolved merge conflicts

yashraj-01 committed:
Nit fix

yashraj-01 committed:
fixed bazel lint

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/forcenetworktype/ForceNetworkTypeViewModel.kt:
```diff
@@ -0,0 +1,31 @@
+import javax.inject.Inject
+
+/**
+ * [ViewModel] for [ForceNetworkTypeFragment]. It populates the recyclerview with a list of
```
I suggest either

```suggestion
 * [ViewModel] for [ForceNetworkTypeFragment]. It populates the RecyclerView with a list of
```

or

```suggestion
 * [ViewModel] for [ForceNetworkTypeFragment]. It populates the recycler view with a list of
```

("recyclerview" isn't a word, but "RecyclerView" is a proper noun versus "recycler view" which spells it out).

Ditto everywhere else.

yashraj-01 replied:
Done. Followed the below one. Thanks.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/forcenetworktype/NetworkTypeItemViewModel.kt:
```diff
@@ -0,0 +1,14 @@
+  val networkType: NetworkConnectionUtil.ConnectionStatus
+) : ObservableViewModel() {
+
+  /** Returns the network type in a human readable format. */
+  fun formatNetworkTypeString(): String = networkType.name.toLowerCase().capitalize()
```
If this is being displayed to the user it ought to use strings instead of formatting the enum constant (to facilitate things like translations & to avoid refactors/enum renames changing the UI).

yashraj-01 replied:
I have modified the approach to use strings instead of enum names. PTAL. Thanks.

BenHenning replied:
This looks much better--thanks!

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/forcenetworktype/ForceNetworkTypeActivity.kt:
```diff
@@ -0,0 +1,28 @@
+import org.oppia.android.app.activity.InjectableAppCompatActivity
+import javax.inject.Inject
+
+/** Activity for Force Network Type. */
```
```suggestion
/** Activity for forcing the network mode for the app. */
```

Prefer being more detailed (& to be judicious about capitalizing words). Ditto for other KDocs in this PR.

yashraj-01 replied:
Done.

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -468,6 +468,7 @@
   <string name="mark_stories_completed_activity_title">Mark Stories Completed</string>
   <string name="mark_topics_completed_activity_title">Mark Topics Completed</string>
   <string name="view_event_logs_activity_title">View Event Logs</string>
+  <string name="force_network_type_activity_title">Force Network Type</string>
```
I wonder if this should be 'Mode' instead of 'Type' since that might be a bit clearer, but I'm not sure.

@rt4914 any thoughts?

BenHenning replied:
I don't have a strong opinion here; resolving to unblock the PR.

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -480,4 +481,5 @@
   <string name="developer_options_force_crash">Force crash</string>
   <string name="all">All</string>
   <string name="mark_completed_uppercase">MARK COMPLETED</string>
+  <string name="network_selected">Network is selected</string>
```
This string name is too generic (keep in mind that these names are globals for the entire app). Can you please update this to be more specific to where the string is being used?

Further, the name should not reflect its exact English contents but rather its semantic meaning (consider "all" vs. "developer_options_modify_lesson_progress_label"; the latter is more specific to the context than the actual contents of the string).

yashraj-01 replied:
Done. PTAL. Thanks.

BenHenning replied:
Much better, thanks!

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/ForceNetworkTypeFragmentTest.kt:
```diff
@@ -0,0 +1,405 @@
+package org.oppia.android.app.devoptions
```
This test should be in the same corresponding package as the component that it's testing. Ditto for all tests in this & future PRs.

yashraj-01 replied:
Refactored. PTAL. Thanks.

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/forcenetworktype/ForceNetworkTypeFragment.kt:
```diff
@@ -0,0 +1,33 @@
+import org.oppia.android.app.fragment.InjectableFragment
+import javax.inject.Inject
+
+/** Fragment to provide functionality to force network type of the app. */
```
```suggestion
/** Fragment to provide functionality to force the network type of the app. */
```

Small nit on grammar.

yashraj-01 replied:
Done.

BenHenning reviewed (changes requested):
Thanks @yashraj-01. Took a pass for my codeowners files (& the ones that they are including changes for).

yashraj-01 committed:
Removed common test deps

yashraj-01 committed:
moved networking test util and its test to testing module

yashraj-01 committed:
bazel nit fix

yashraj-01 committed:
Merge branch 'prod-and-debug-networkconnectionutil' into ui-force-network-type

yashraj-01 committed:
Moved activity and fragment tests to forcenetworktype package

yashraj-01 committed:
Nit fixes

yashraj-01 committed:
Used string instead of enum names

yashraj-01 committed:
Fixed failing static checks

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
Fixed failing test

yashraj-01 committed:
merged develop and resolved merge conflicts

yashraj-01 committed:
nit fix

yashraj-01 committed:
merged base and resolved merge conflicts

yashraj-01 committed:
Removed redundant kdoc exemption

BenHenning commented:
Hi. Will need to look at this tomorrow.

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

BenHenning reviewed (commented):
Thanks @yashraj-01. LGTM, but holding off on approving until the base PR is merged.

Also, @yashraj-01 could you please fix the corresponding static analysis & CI checks after merging the base PR and before assigning this back?

yashraj-01 committed:
Made NetworkConnectionDebugUtil optional

yashraj-01 committed:
Fixed tests

yashraj-01 committed:
fixed bazel build and added TODOs

yashraj-01 committed:
Added Kdocs and test file exemptions

yashraj-01 committed:
Merge remote-tracking branch 'upstream/develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
Added network util module in test file

yashraj-01 committed:
merged base and resolved merge conflicts

yashraj-01 committed:
Added debug util module

yashraj-01 commented:
> Thanks @yashraj-01. LGTM, but holding off on approving until the base PR is merged.
> 
> Also, @yashraj-01 could you please fix the corresponding static analysis & CI checks after merging the base PR and before assigning this back?

I have updated the PR with the base PR and also fixed the static checks. PTAL. Thanks.

yashraj-01 committed:
fixed bazel failure

BenHenning commented on app/src/main/java/org/oppia/android/app/devoptions/forcenetworktype/ForceNetworkTypeFragmentPresenter.kt:
```diff
@@ -73,11 +75,15 @@ class ForceNetworkTypeFragmentPresenter @Inject constructor(
-    binding.networkTypeLayout.setOnClickListener {
-      networkConnectionUtil.setCurrentConnectionStatus(model.networkType)
-      bindingAdapter.notifyDataSetChanged()
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
```
This won't work on older versions of Android, and I'm not sure if this is covered under desugaring support. Suggest using the imperative if-check using ``isPresent()`` instead.

yashraj-01 replied:
I think Optional is covered under desugaring support [reference](https://developer.android.com/studio/write/java8-support#library-desugaring). But I think it requires building the app with Gradle plugin version 4.0.0+ and the current version is 3.6.1. And I'm not sure what are its requirements for Bazel. Also, Optional is introduced in API level 24 so all its methods are accessible in version > N only (even the get method), so we can't use an if statement here (correct me if I'm wrong). I have already had a TODO for it which I have added here.

BenHenning replied:
Let's use Guava's Optional, instead. That should let us avoid the issue.

yashraj-01 replied:
Thanks @BenHenning. This solved the issue. Updated the PR to use Guava's Optional instead of Java8.

BenHenning reviewed (commented):
Thanks @yashraj-01. Just had one follow-up comment. Otherwise, the passing CI checks LGTM. Please reassign once the last comment is resolved & #3512 is merged in and this PR updated.

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
renamed connection status

yashraj-01 committed:
Nit fix

yashraj-01 committed:
introduced oppia_debug_visibility

yashraj-01 committed:
merged base and resolved merge conflicts

yashraj-01 committed:
added TODO

yashraj-01 committed:
fixed incorrect connection status

BenHenning reviewed (commented):
Thanks @yashraj-01. Had one follow-up.

yashraj-01 committed:
used Guava's optional instead of java8

yashraj-01 committed:
Merge branch 'develop' into prod-and-debug-networkconnectionutil

yashraj-01 committed:
Merge branch 'prod-and-debug-networkconnectionutil' into ui-force-network-type

yashraj-01 committed:
merged develop and resolved merge conflicts

BenHenning reviewed (approved):
Thanks @yashraj-01. This LGTM!

anandwana001 reviewed (approved):
LGTM for code owner files.
